rllib/agents/slateq/slateq.py

```diff
@@ -0,0 +1,220 @@
+"""
+SlateQ (Reinforcement Learning for Recommendation)
+==================================================
+
+This file defines the trainer class for the SlateQ algorithm from the
+`"Reinforcement Learning for Slate-based Recommender Systems: A Tractable
+Decomposition and Practical Methodology" <https://arxiv.org/abs/1905.12767>`_
+paper.
+
+See `slateq_torch_policy.py` for the definition of the policy. Currently, only
+PyTorch is supported. The algorithm is written and tested for Google's RecSim
+environment (https://github.com/google-research/recsim).
+"""
+
+import logging
+from typing import List, Type
+
+from ray.rllib.agents.slateq.slateq_torch_policy import SlateQTorchPolicy
+from ray.rllib.agents.trainer import with_common_config
+from ray.rllib.agents.trainer_template import build_trainer
+from ray.rllib.evaluation.worker_set import WorkerSet
+from ray.rllib.examples.policy.random_policy import RandomPolicy
+from ray.rllib.execution.concurrency_ops import Concurrently
+from ray.rllib.execution.metric_ops import StandardMetricsReporting
+from ray.rllib.execution.replay_buffer import LocalReplayBuffer
+from ray.rllib.execution.replay_ops import Replay, StoreToReplayBuffer
+from ray.rllib.execution.rollout_ops import ParallelRollouts
+from ray.rllib.execution.train_ops import TrainOneStep
+from ray.rllib.policy.policy import Policy
+from ray.rllib.utils.typing import TrainerConfigDict
+from ray.util.iter import LocalIterator
+
+logger = logging.getLogger(__name__)
+
+# Defines all SlateQ strategies implemented.
+ALL_SLATEQ_STRATEGIES = [
+    # RANDOM: Randomly select documents for slates.
+    "RANDOM",
+    # MYOP: Select documents that maximize user click probabilities. This is
+    # a myopic strategy and ignores long term rewards. This is equivalent to
+    # setting a zero discount rate for future rewards.
+    "MYOP",
+    # SARSA: Use the SlateQ SARSA learning algorithm.
+    "SARSA",
+    # QL: Use the SlateQ Q-learning algorithm.
+    "QL",
+]
+
+# yapf: disable
+# __sphinx_doc_begin__
+DEFAULT_CONFIG = with_common_config({
+    # === Model ===
+    # Dense-layer setup for each the advantage branch and the value branch
+    # in a dueling architecture.
+    "hiddens": [256, 64, 16],
+
+    # set batchmode
+    "batch_mode": "complete_episodes",
```

Comment:
Does this have to be enforced? If yes, we should add a check to `validate_config`.

Reply:
Only SARSA training needs this. I'll add a check in validate_config.